File: 3474.lexicographically-smallest-generated-string.cpp
#
# @lc app=leetcode id=3474 lang=cpp
#
# [3474] Lexicographically Smallest Generated String
#

# @lc code=start
class Solution {
public:
    std::string generateString(std::string str1, std::string str2) {
        size_t n = str1.size();
        size_t m = str2.size();
        size_t L = n + m - 1;
        std::string word(L, '\0');
        // Step 2: Propagate 'T' constraints
        for (size_t i = 0; i < n; ++i) {
            if (str1[i] != 'T') continue;
            for (size_t j = 0; j < m; ++j) {
                size_t pos = i + j;
                char req = str2[j];
                if (word[pos] != '\0' && word[pos] != req) {
                    return "";
                }
                word[pos] = req;
            }
        }
        // Step 3: Check forced 'F' matches
        for (size_t i = 0; i < n; ++i) {
            if (str1[i] != 'F') continue;
            bool mismatch = false;
            bool has_free = false;
            for (size_t j = 0; j < m; ++j) {
                size_t pos = i + j;
                if (word[pos] == '\0') {
                    has_free = true;
                    continue;
                }
                if (word[pos] != str2[j]) {
                    mismatch = true;
                }
            }
            if (!mismatch && !has_free) {
                return "";
            }
        }
        // Step 4: Greedily fill free positions
        for (size_t k = 0; k < L; ++k) {
            if (word[k] != '\0') continue;
            bool forbidden[26] = {};
            size_t i_start = std::max(static_cast<size_t>(0), k + 1 - m);
            size_t i_max = std::min(n - 1, k);
            for (size_t i = i_start; i <= i_max; ++i) {
                if (str1[i] != 'F') continue;
                size_t j = k - i;
                // Check prefix match
                bool prefix_ok = true;
                for (size_t p = 0; p < j; ++p) {
                    if (word[i + p] != str2[p]) {
                        prefix_ok = false;
                        break;
                    }
                }
                if (!prefix_ok) continue;
                // Check future fully fixed and match
                bool future_ok = true;
                for (size_t p = j + 1; p < m; ++p) {
                    size_t pos = i + p;
                    if (word[pos] == '\0' || word[pos] != str2[p]) {
                        future_ok = false;
                        break;
                    }
                }
                if (future_ok) {
                    forbidden[str2[j] - 'a'] = true;
                }
            }
            // Choose smallest allowed
            bool found = false;
            for (char c = 'a'; c <= 'z'; ++c) {
                if (!forbidden[c - 'a']) {
                    word[k] = c;
                    found = true;
                    break;
                }
            }
            if (!found) {
                return "";
            }
        }
        // Step 5: Verify all constraints
        for (size_t i = 0; i < n; ++i) {
            bool is_match = true;
            for (size_t j = 0; j < m; ++j) {
                if (word[i + j] != str2[j]) {
                    is_match = false;
                    break;
                }
            }
            if ((str1[i] == 'T' && !is_match) || (str1[i] == 'F' && is_match)) {
                return "";
            }
        }
        return word;
    }
};
# @lc code=end

File: 3474.lexicographically-smallest-generated-string.java
#
# @lc app=leetcode id=3474 lang=java
#
# [3474] Lexicographically Smallest Generated String
#

# @lc code=start
class Solution {
    public String generateString(String str1, String str2) {
        int n = str1.length(), m = str2.length();
        int len = n + m - 1;
        char[] res = new char[len];
        Arrays.fill(res, '?');
        // Step 1: place all required str2s
        for (int i = 0; i < n; ++i) {
            if (str1.charAt(i) == 'T') {
                for (int j = 0; j < m; ++j) {
                    int pos = i + j;
                    if (res[pos] == '?') {
                        res[pos] = str2.charAt(j);
                    } else if (res[pos] != str2.charAt(j)) {
                        return ""; // Conflict
                    }
                }
            }
        }
        // Step 2: mark forbidden positions for F
        for (int i = 0; i < len; ++i) {
            if (res[i] == '?') {
                res[i] = 'a';
            }
        }
        // Step 3: check F constraints and try to fix if needed
        for (int i = 0; i < n; ++i) {
            if (str1.charAt(i) == 'F') {
                boolean match = true;
                for (int j = 0; j < m; ++j) {
                    if (res[i + j] != str2.charAt(j)) {
                        match = false;
                        break;
                    }
                }
                if (match) {
                    // Try to break lex smallest
                    boolean changed = false;
                    for (int j = m - 1; j >= 0 && !changed; --j) {
                        int pos = i + j;
                        for (char c = (char) (str2.charAt(j) + 1); c <= 'z'; ++c) {
                            if (res[pos] == str2.charAt(j)) {
                                // Check if changing to c is valid (doesn't break T's)
                                boolean ok = true;
                                for (int k = 0; k < n; ++k) {
                                    if (str1.charAt(k) == 'T') {
                                        int start = k, end = k + m - 1;
                                        if (pos >= start && pos <= end) {
                                            if (str2.charAt(pos - start) != c) {
                                                ok = false;
                                                break;
                                            }
                                        }
                                    }
                                }
                                if (ok) {
                                    res[pos] = c;
                                    changed = true;
                                    break;
                                }
                            }
                        }
                    }
                    if (!changed) return "";
                }
            }
        }
        // Step 4: Final verification of all constraints
        for (int i = 0; i < n; ++i) {
            boolean match = true;
            for (int j = 0; j < m; ++j) {
                if (res[i + j] != str2.charAt(j)) {
                    match = false;
                    break;
                }
            }
            if ((str1.charAt(i) == 'T' && !match) || (str1.charAt(i) == 'F' && match)) {
                return "";
            }
        }
        return new String(res);
    }
}
# @lc code=end